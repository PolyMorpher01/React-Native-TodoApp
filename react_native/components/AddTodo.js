import React, { Component } from 'react';
import { TextInput, StyleSheet, View, TouchableOpacity, Text } from 'react-native';

import TodoService from '../realm/service/todoService'

class AddTodo extends Component {
    constructor(props) {
        super(props);
        this.state = {
            newTask: '',
            note: '',
            showBottomView: false,
        }
    }

    saveTodo() {
        TodoService.save(this.state.newTask);
        this.setState({
            newTask: ''
        });
        navigator.navigate
    }



    renderBottomView() {
        if (this.state.showBottomView) {
            return (
                <View>
                    <TextInput style={styles.textInput}
                        placeholder='Add a Note'
                        multiline={true}
                        numberOfLines={3}
                        maxLength={40}
                        value={this.state.note}
                        onChangeText={(text) => this.setState({ note: text })} />

                    <TouchableOpacity style={styles.button}
                    // onPress={() => this.onSaveClicked()}      
                    >
                        <Text style={styles.buttonText}>Save</Text>
                    </TouchableOpacity>
                </View>
            );
        }
        return null;
    }



    render() {
        return (
            <View style={styles.container}
            // onBlur={() => this.setState({ showBottomView: false })}
            >
                <TextInput style={styles.textInput}
                    onSubmitEditing={() => { this.saveTodo() }}
                    // onFocus={() => this.setState({ showBottomView: true })}
                    placeholder='Add a todo'
                    value={this.state.newTask}
                    onChangeText={(text) => this.setState({ newTask: text })} />
                {this.renderBottomView()}
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        margin: 10
    },
    textInput: {
        padding: 10,
        fontSize: 18,
        borderWidth: 1,
        borderColor: '#eee',
        borderRadius: 8,
        backgroundColor: '#fff',
        marginTop: 10,
        marginBottom: 10,
    },
    button: {
        height: 44,
        backgroundColor: '#03A9F4',
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: 8,
        margin: 1,
        marginTop: 10,
        marginBottom: 10,
    },
    buttonText: {
        fontSize: 18,
        color: 'white',
        textAlign: 'center',
        fontWeight: 'bold',
        marginTop: 10,
        marginBottom: 10,
    },
});

export default AddTodo;