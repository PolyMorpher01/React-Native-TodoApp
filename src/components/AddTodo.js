import React, { Component } from 'react';
import { TextInput, StyleSheet, View, TouchableOpacity, Text } from 'react-native';

import TodoService from '../realm/service/todoService';
import Color from '../constants/Color'

class AddTodo extends Component {

    state = {
        newTask: '',
        note: '',
        showBottomView: false
    }

    saveTodo = () => {
        const newTask = this.state.newTask.trim();
        if (newTask === '') {
            return;
        }
        let todoList = TodoService.save(newTask);
        this.setState({
            newTask: ''
        });

        this.props.updateTodoList(todoList);
    }



    render() {
        return (
            <View style={styles.container}>
                <TextInput style={styles.textInput}
                    onSubmitEditing={this.saveTodo}
                    placeholder='Add a todo'
                    value={this.state.newTask}
                    onChangeText={(text) => this.setState({ newTask: text })} />
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        margin: 10
    },
    textInput: {
        padding: 10,
        fontSize: 18,
        borderWidth: 1,
        borderColor: Color.VERY_LIGHT_GREY,
        borderRadius: 8,
        backgroundColor: Color.WHITE,
        marginTop: 10,
        marginBottom: 10,
    },
    button: {
        height: 44,
        backgroundColor: Color.APP_THEME,
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: 8,
        margin: 1,
        marginTop: 10,
        marginBottom: 10,
    },
    buttonText: {
        fontSize: 18,
        color: 'white',
        textAlign: 'center',
        fontWeight: 'bold',
        marginTop: 10,
        marginBottom: 10,
    },
});

export default AddTodo;