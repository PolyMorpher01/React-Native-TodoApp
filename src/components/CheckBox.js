import React, { Component } from 'react';
import Icon from 'react-native-vector-icons/MaterialIcons';

import Color from '../constants/Color'

class CheckBox extends Component {
    constructor(props) {
        super(props);
        this.state = {
            data: this.props.data
        };
    }

    render() {
        let iconName = this.state.data.isCompleted ? 'check-box' : 'check-box-outline-blank';

        return (
            <Icon.Button
                data={this.state.data}
                name={iconName}
                backgroundColor='rgba(0,0,0,0)'
                color={Color.BLACK}
                underlayColor='rgba(0,0,0,0)'
                activeOpacity={0.6}
                onPress={this.props.onCheckBoxPressed}
            >
            </Icon.Button>
        );
    }
}

export default CheckBox;