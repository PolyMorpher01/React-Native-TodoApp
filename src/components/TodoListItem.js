import React, { Component } from 'react';
import { TouchableHighlight, View, Text, StyleSheet } from 'react-native';
import { Menu, MenuOptions, MenuOption, MenuTrigger } from 'react-native-popup-menu';

import CheckBox from './CheckBox';
import TodoService from '../realm/service/todoService';
import Color from '../constants/Color'


class TodoListItem extends Component {
    constructor(props) {
        super(props);
        this.state = {
            todo: this.props.todo
        }
    }

    changeCompletedStatus = () => {
        var todo = this.state.todo;
        TodoService.update(() => {
            todo.isCompleted = !todo.isCompleted;
        });
        this.setState({
            todo: todo
        });
    }

    renderMenu = (todo) => {
        return (
            <Menu>
                <MenuTrigger text='...' style={{ padding: 8 }} />
                <MenuOptions>
                    <MenuOption onSelect={() => this.props.onShowDetail(todo)} text='View Detail' style={styles.menuOption} />
                    <MenuOption onSelect={() => this.props.onDeleteTodo(todo)} text='Delete' style={styles.menuOption} />
                </MenuOptions>
            </Menu>
        );
    }

    render() {
        let todo = this.state.todo;
        let color = todo.isCompleted ? Color.LIGHT_GREY : Color.BLACK;
        let textDecorationLine = todo.isCompleted ? 'line-through' : 'none';

        return (
            <TouchableHighlight style={styles.itemContainer}
                underlayColor={Color.LIST_ITEM_UDERLAY}
                onPress={this.changeCompletedStatus}
            >
                <View style={styles.item}>
                    <CheckBox data={todo} style={styles.checkBox} onCheckBoxPressed={this.changeCompletedStatus} />
                    <Text style={[styles.itemText, { color: color, textDecorationLine: textDecorationLine, }]}>{todo.task}</Text>
                    {this.renderMenu(todo)}
                </View>
            </TouchableHighlight>
        )
    }
}


const styles = StyleSheet.create({
    item: {
        padding: 10,
        fontSize: 18,
        flex: 1,
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'space-between',
        flex: 1
    },
    itemContainer: {
        borderBottomWidth: 1,
        borderBottomColor: Color.GREY,
        flex: 1,
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'space-between',
        padding: 4
    },
    checkBox: {
        flex: 1
    },
    itemText: {
        fontSize: 18,
        flex: 1
    },
    menuOption: {
        padding: 10
    },
});

export default TodoListItem;
