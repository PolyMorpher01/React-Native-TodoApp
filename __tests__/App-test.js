/**
 * @format
 */

import 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { MenuProvider } from 'react-native-popup-menu';

import TodoService from '../src/realm/service/todoService'
import RealmUtils from '../src/realm/utils';
import AddTodo from '../src/components/AddTodo';
import CheckBox from '../src/components/CheckBox';
import TodoListItem from '../src/components/TodoListItem';


function initalizeDatabase() {
    const tasks = ["one", "two", "three"];
    TodoService.saveMultiple(tasks);
}

function clearDatabase() {
    TodoService.deleteAll();
}

beforeAll(() => {
    return initalizeDatabase();
});


afterAll(() => {
    return clearDatabase();
})

describe('Realm Tasks', () => {
    test('Todo list exists', () => {
        const todoList = TodoService.findall();
        expect(todoList).toBeTruthy();
    });

    test('Task one exists', () => {
        const id = 1;
        const todo = TodoService.findById(id);
        expect(todo.task).toBe("one");
    })

    test('Task updates', () => {
        const id = 1;
        const todo = TodoService.findById(id);
        TodoService.update(() => {
            todo.task = "one updated";
        })
        expect(todo.task).toBe("one updated");
    })

    test('Task is deleted', () => {
        const id = 1;
        const todo = TodoService.findById(id);
        TodoService.delete(todo);
        const todoDeleted = TodoService.findById(id);
        expect(todoDeleted).toBeFalsy();
    })

    test('Realm primary key generation', () => {
        const todoList = TodoService.findall();
        const id = RealmUtils.getMaxIdForPrimaryKey(todoList);
        const isNumber = Number.isInteger(id);
        expect(id).toBeTruthy();
        expect(id).toBeGreaterThanOrEqual(0);
        expect(isNumber).toBeTruthy();
    })
});

describe('Component renders correctly', () => {

    it('AddTodo', () => {
        const instance = renderer.create(
            <AddTodo />);
        expect(instance.toJSON()).toMatchSnapshot();
    });

    it('CheckBox', () => {
        const todo = TodoService.findById(2);
        const instance = renderer.create(
            <CheckBox
                data={todo}
            />);
        expect(instance.toJSON()).toMatchSnapshot();
    });

    it('Todo List Item', () => {
        const todo = TodoService.findById(2);
        const instance = renderer.create(
            //Menu Component inside TodoListItem requires a MenuProvider as root component
            <MenuProvider>
                <TodoListItem
                    todo={todo}
                />
            </MenuProvider>
        );
        expect(instance.toJSON()).toMatchSnapshot();
    });
})
