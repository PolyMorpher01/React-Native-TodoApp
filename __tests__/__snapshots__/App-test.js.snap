// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component renders correctly AddTodo 1`] = `
<View
  style={
    Object {
      "margin": 10,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Add a todo"
    rejectResponderTermination={true}
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderColor": "#E0E0E0",
        "borderRadius": 8,
        "borderWidth": 1,
        "fontSize": 18,
        "marginBottom": 10,
        "marginTop": 10,
        "padding": 10,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;

exports[`Component renders correctly CheckBox 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "transparent",
        "borderRadius": 5,
      },
    ]
  }
>
  <View
    data={
      Todo {
        "id": 2,
        "isCompleted": false,
        "task": "two",
      }
    }
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "padding": 8,
        },
        Object {
          "backgroundColor": "transparent",
          "borderRadius": 5,
        },
        undefined,
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 20,
          },
          Object {
            "marginRight": 10,
          },
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`Component renders correctly Todo List Item 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderBottomColor": "#333333",
          "borderBottomWidth": 1,
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 4,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "fontSize": 18,
            "justifyContent": "space-between",
            "padding": 10,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "backgroundColor": "transparent",
                "borderRadius": 5,
              },
            ]
          }
        >
          <View
            data={
              Todo {
                "id": 2,
                "isCompleted": false,
                "task": "two",
              }
            }
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                  "padding": 8,
                },
                Object {
                  "backgroundColor": "transparent",
                  "borderRadius": 5,
                },
                undefined,
              ]
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontSize": 20,
                  },
                  Object {
                    "marginRight": 10,
                  },
                  Object {
                    "fontFamily": "Material Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
        <Text
          style={
            Array [
              Object {
                "flex": 1,
                "fontSize": 18,
              },
              Object {
                "color": "#000000",
                "textDecorationLine": "none",
              },
            ]
          }
        >
          two
        </Text>
        <View>
          <View
            collapsable={false}
          >
            <View
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={null}
            >
              <View
                ctx={
                  Object {
                    "menuActions": Object {
                      "_getOpenedMenu": [Function],
                      "_notify": [Function],
                      "closeMenu": [Function],
                      "isMenuOpen": [Function],
                      "openMenu": [Function],
                      "toggleMenu": [Function],
                    },
                    "menuRegistry": Object {
                      "getAll": [Function],
                      "getMenu": [Function],
                      "setOptionsCustomStyles": [Function],
                      "subscribe": [Function],
                      "unsubscribe": [Function],
                      "updateLayoutInfo": [Function],
                    },
                  }
                }
                style={
                  Array [
                    undefined,
                    Object {
                      "padding": 8,
                    },
                  ]
                }
              >
                <Text>
                  ...
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={false}
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </RCTSafeAreaView>
</View>
`;
